### PR TITLE
add search/play playlist feature

### PR DESCRIPTION
**Pull request by dpnick**
## Description

Add two commands to "Apple Music" plugin:
- find and start any of your playlists
- search and play a track from your library

- Bug fix / improvement for somebody else's extension. Extension author: @Rawnly 

## Screencast

Start Playlist

https://user-images.githubusercontent.com/23317360/151350673-82043707-4feb-4794-86b3-4eca4459bfd0.mov

Search & Play track

https://user-images.githubusercontent.com/23317360/151350729-196f7a80-5cb7-4b7e-a4c7-e46e9f04d999.mov

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)

### CODE REVIEW

**Comment by pernielsentikaer** [comment]
Hi @dpnick 👋 

Thanks for your contribution 💪 It's always a pleasure when extensions are getting more features 💯 

@Rawnly can you try it out and give some feedback?

**Comment by dpnick** [comment]
Hi @pernielsentikaer 👋

Thanks, happy to contribute to this nice project! 

Do not hesitate if you have any feedback! 

PS: have an extra look to the "play album" feature. Applescript doesn't allow to directly play every tracks from an album in a row. I had to create a temporary playlist and then to play it. It is maybe too "hacky" ?

**Comment by pernielsentikaer** [comment]
@Rawnly are you still around?

**Comment by Rawnly** [comment]
Sure, sorry guys, I had a very busy week, I'll take a look in the weekend ✌️

**Comment by pernielsentikaer** [comment]
> Sure, sorry guys, I had a very busy week, I'll take a look in the weekend ✌️

No problem at all, and it's only if you have time, no rush at all — I'm just happy that you have the time 😍

File: extensions/music/src/play-library-album.tsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  const [albums, setAlbums] = useState<Album[] | null>();
+  const { pop } = useNavigation();
+
+  const onSearch = async (next: string) => {
+    setAlbums(null);
+    if (!next || next?.length < 1) {
+      setAlbums([]);
+      return;
+    }
+    const raw = await seachForAlbum(next)();
```
Comment (Rawnly) [inline extensions/music/src/play-library-album.tsx:18]: I would write this in a more functional way, following `fp-ts` docs. 
But even like this is ok for me, I can't force you to write code in a weird way 😄 

Something like this should be fine.. 👀  (check types)
```ts
const result: Album[] = await pipe(
	next,
	searchForAlbum,
	TE.matchW(
		() => showToast(ToastStyle.Failure, 'Could not get albums'),
		albums => pipe(
			albums,
			O.fromNullable,
			O.matchW(
				() => [],
				A.map(album => parseResult<Album>(album))
			)
		)
	),
)()
```

File: extensions/music/src/play-library-track.tsx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+export default function PlayLibraryTrack() {
+  const [tracks, setTracks] = useState<Track[] | null>();
+  const { pop } = useNavigation();
+
+  const onSearch = async (next: string) => {
+    if (!next || next?.length < 1) {
+      setTracks([]);
+      return;
+    }
+    const raw = await seachForTrack(next)();
```
Comment (Rawnly) [inline extensions/music/src/play-library-track.tsx:17]: Same as `play-library-album`

**Review by Rawnly — COMMENTED** [review]
Just some thoughts on `Promise/TaskEither` flow. If you want to refactor is ok, otherwise i can always open a new PR in the future 😃

**Comment by Rawnly** [comment]
> PS: @Rawnly have an extra look to the "play album" feature. Applescript doesn't allow to directly play every tracks from an album in a row. I had to create a temporary playlist and then to play it. It is maybe too "hacky" ?

This is ok for me, a bit tricky but it works :)

File: extensions/music/src/play-library-album.tsx (hunk 1 above)
Comment (Rawnly) [reply extensions/music/src/play-library-album.tsx:18]: If you need any help or have some doubts you can contact me on Slack

**Comment by pernielsentikaer** [comment]
Thanks for the feedback @Rawnly 💯 💪

**Comment by electr1fy0** [comment]
Is a command to search the entire Apple Music exactly like the Spotify Extension possible?

**Comment by pernielsentikaer** [comment]
Can one of you two answer the question above? @Rawnly / @dpnick

**Comment by pernielsentikaer** [comment]
Are you still available to check on the corrections @dpnick 😄

**Comment by Rawnly** [comment]
> Can one of you two answer the question above? @Rawnly / @dpnick

I don't think that this can be done right now, not via the `apple-script` commands. I don't know if apple music provides some kind of api to play with.

Btw, I'll take a look asap.

**Comment by mattisssa** [comment]
Yeah, there is API for this, just makes a development a bit complexier 😉

**Comment by Rawnly** [comment]
> Yeah, there is API for this, just makes a development a bit complex 😉

I took a rapid look at The [Apple Music Api docs](https://developer.apple.com/documentation/applemusicapi), my only problem is that for testing the integration you [need a token](https://developer.apple.com/documentation/applemusicapi/getting_keys_and_creating_tokens) and to get the token you must be enrolled in the **Apple Developer Program** 🤔 (90$/yr for non "swifty" readers)

I was enrolled a few years ago but I'm not anymore (it was just an hobby) 😅

**Comment by dpnick** [comment]
Sorry everyone, I have been really busy lately!

@pernielsentikaer about the corrections, I will have a look asap!

About the "search through all Apple Music" feature, I think we should do it on another PR 😅

**Comment by Rawnly** [comment]
> About the "search through all Apple Music" feature, I think we should do it on another PR 😅

Uh yeah that was implicit i think 😄

**Comment by pernielsentikaer** [comment]
Awesome @dpnick and I'm just happy that you have the time to check it ❤️

**Comment by dpnick** [comment]
Oh yes sorry for the delay, I have been really busy these days, I will do it today!

**Comment by dpnick** [comment]
Ok @Rawnly , I tried to implement the suggested changes using `fp-ts` functional way. Since I didn't know really well this library, please check the code 🙂

**Comment by pernielsentikaer** [comment]
Awesome @dpnick – thanks for taking your time 💪 

@Rawnly do you have time to check it? 🔥

**Comment by Rawnly** [comment]
Sure, I'll take a look asap

**Comment by pernielsentikaer** [comment]
Thanks! 😄

**Review by Rawnly — APPROVED** [review]

**Comment by Rawnly** [comment]
Sorry for being super late. Just reviewed! Everything is ok for me! @pernielsentikaer

**Pull request merged**